docs/RFCS/monitoring_queries_and_sessions.md

```diff
@@ -0,0 +1,219 @@
+
+Since some DB deployments can be shared between multiple client applications, some level of
+functionality to filter sessions and queries by client application names, both in the admin
+UI and in the CLI, would also be useful. And since not all DBAs are familiar with SQL,
+a non-SQL interface to monitor queries and sessions would be an asset.
+
+# Detailed design
+
+The inspiration behind some of the design comes from
+[this PR from Raphael](https://github.com/cockroachdb/cockroach/pull/10317) which implemented
+node-local session registries exposed via a virtual table.
+
+## Thread-safe session struct
+
+The session struct is currently not thread-safe; since all operations on it
+currently happen in one thread. Making it thread-safe involves making a nested struct
+inside it (called `mu`), similar to how the `Txn` struct is structured. There
+would be an `RWMutex` on this nested `mu` struct.
+
+These fields would be moved from the main `Session` struct, to the `mu` sub-struct:
+- User (name of user)
+- Remote address (IP, port)
+- Application Name
+- TxnState
+- Eventually: any contexts/object references useful for cancellation
+
+These two new fields will be added to `mu`:
+- Start timestamp
+- Active query (reference to Query object defined below, can be nil)
+
+The active query reference will contain a reference to a non-parallel query _if one is running_.
+If a set of parallelizable queries are running, they will all be in `parallelizeQueue`. Inspecting
+both the active query reference and the `parallelizeQueue` will be required to know all queries
+in flight.
+
+Write locks on this struct will be acquired when:
+- A new session is created (all fields are populated)
+- The application name is reset (in `resetApplicationName()`)
+- A query begins or ends execution (to updated the active query reference / parallelize queue)
+- txnState.txn is updated in `executor.execRequest`. 
+
+## Node-local session registry
+
+The session registry will be an in-memory slice that stores references to all
+session objects on that node. Addition and deletion of new
+sessions to this slice would require acquiring a mutex lock.
+
+This addition/deletion into the slice can be made in `NewSession(...)` and `Session.Finish(...)`
+in `pkg/sql/session.go`.
+
+## New Query struct
+
+A new struct, `Query`, would be made that contains the below fields:
+
+- Query string (output of `planToString()`)
+- Start timestamp
+- Reference to root `planNode` (which it could effectively replace)
+- Eventually: any contexts/object references useful for cancellation 
+
+The parallelizeQueue will be updated to take Query objects instead of planNodes. Non-parallelized
+queries will be directly referenced via a pointer in `Session`.
+
+## RPC endpoint in status server to return local queries and sessions
+
+Two new RPC endpoints would be made: `LocalSessions` and `LocalQueries`. These
```

Comment:
We might be able to use LocalSessions for querying session data and use that to get to the query data without needing the extra LocalQueries.